worker/rhpv2.go

```diff
@@ -262,120 +266,386 @@ func (w *worker) FetchSignedRevision(ctx context.Context, hostIP string, hostKey
-		}
+	// sort in descending order so that we can use 'range'
+	sort.Slice(indices, func(i, j int) bool {
+		return indices[i] > indices[j]
+	})
 
-		// extract the revision
-		rev := rhpv2.ContractRevision{
-			Revision:   revision,
-			Signatures: [2]types.TransactionSignature{sigs[0], sigs[1]},
-		}
+	// decide on the batch size, defaults to ~20mib of sector data but for old
+	// hosts we use a much smaller batch size to ensure we nibble away at the
+	// problem rather than outright failing or timing out
+	batchSize := int(batchSizeDeleteSectors)
+	if build.VersionCmp(settings.Version, "1.6.0") < 0 {
+		batchSize = 100
+	}
 
-		// execute settings RPC
-		var settingsResp rhpv2.RPCSettingsResponse
-		if err := t.Call(rhpv2.RPCSettingsID, nil, &settingsResp); err != nil {
-			return err
+	// split the indices into batches
+	var batches [][]uint64
+	for {
+		if len(indices) < batchSize {
+			batchSize = len(indices)
 		}
-		var settings rhpv2.HostSettings
-		if err := json.Unmarshal(settingsResp.Settings, &settings); err != nil {
-			return fmt.Errorf("couldn't unmarshal json: %w", err)
+		batches = append(batches, indices[:batchSize])
+		indices = indices[batchSize:]
+		if len(indices) == 0 {
+			break
 		}
+	}
 
-		// download all roots
-		numsectors := rev.NumSectors()
-		for offset := uint64(0); offset < numsectors; {
-			n := batchSizeFetchSectors
-			if offset+n > numsectors {
-				n = numsectors - offset
+	// derive the renter key
+	renterKey := w.deriveRenterKey(rev.HostKey())
+
+	// range over the batches and delete the sectors batch per batch
+	for i, batch := range batches {
+		if err = func() error {
+			var cost types.Currency
+			start := time.Now()
+			w.logger.Debugw(fmt.Sprintf("starting batch %d/%d of size %d", i+1, len(batches), len(batch)))
+			defer func() {
+				w.logger.Debugw(fmt.Sprintf("processing batch %d/%d of size %d took %v", i+1, len(batches), len(batch), time.Since(start)), "cost", cost)
+			}()
+
+			numSectors := rev.NumSectors()
+
+			// build a set of actions that move the sectors we want to delete
+			// towards the end of the contract, preparing them to be trimmed off
+			var actions []rhpv2.RPCWriteAction
+			cIndex := numSectors - 1
+			for _, rIndex := range batch {
+				if cIndex != rIndex {
+					actions = append(actions, rhpv2.RPCWriteAction{
+						Type: rhpv2.RPCWriteActionSwap,
+						A:    uint64(cIndex),
+						B:    uint64(rIndex),
+					})
+				}
+				cIndex--
 			}
+			actions = append(actions, rhpv2.RPCWriteAction{
+				Type: rhpv2.RPCWriteActionTrim,
+				A:    uint64(len(batch)),
+			})
 
 			// check funds
-			price := rhpv2.RPCSectorRootsCost(settings, n)
-			if rev.RenterFunds().Cmp(price) < 0 {
+			proofSize := uint64(len(batch)) * 2 * uint64(bits.Len64(numSectors)) * 32
+			if proofSize < 4096 {
+				proofSize = 4096
+			}
+
+			cost = settings.BaseRPCPrice.Add(settings.DownloadBandwidthPrice.Mul64(proofSize))
+			cost = cost.Mul64(125).Div64(100)
```

Comment:
More magic without comment. Are we going to follow up with exact calculations once core is updated?